`SELECT * REPLACE` isn't currently handled

Hi David,

https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#select_replace

sqlvalidator 0.0.18:

```diff
- SELECT * REPLACE('bar' AS foo)
+ SELECT * REPLACE
```

A quick unit test causes
```
self = <Alias: <Column: '*'> as=False 'REPLACE'>

    def __str__(self):
>       return str(self.value)
E       AttributeError: 'Alias' object has no attribute 'value'

../../sqlvalidator/grammar/sql.py:239: AttributeError
```
when trying to print the expression

--- a/sqlvalidator/grammar/lexer.py
+++ b/sqlvalidator/grammar/lexer.py
@@ -902,6 +902,10 @@ class ExpressionParser:
                 raise ParsingError("expected '('")
             argument_tokens = get_tokens_until_closing_parenthesis(tokens)
             arguments = ExpressionListParser.parse(iter(argument_tokens))
+            for arg in arguments:
+                assert (
+                    isinstance(arg, Alias) and arg.with_as is True
+                ), "SELECT * REPLACE arguments must be alias with AS"
             expression = ReplaceClause(expression, arguments)
             next_token = next(tokens, None)
 

--- a/sqlvalidator/grammar/lexer.py
+++ b/sqlvalidator/grammar/lexer.py
@@ -35,6 +35,7 @@ from sqlvalidator.grammar.sql import (
     OrderByClause,
     OrderByItem,
     Parenthesis,
+    ReplaceClause,
     SelectStatement,
     String,
     Table,
@@ -893,6 +894,15 @@ class ExpressionParser:
             argument_tokens = get_tokens_until_closing_parenthesis(tokens)
             arguments = ExpressionListParser.parse(iter(argument_tokens))
             expression = ExceptClause(expression, arguments)
+            next_token = next(tokens, None)
+
+        if lower(next_token) == "replace":
+            opening_parenthesis = next(tokens, None)
+            if opening_parenthesis != "(":
+                raise ParsingError("expected '('")
+            argument_tokens = get_tokens_until_closing_parenthesis(tokens)
+            arguments = ExpressionListParser.parse(iter(argument_tokens))
+            expression = ReplaceClause(expression, arguments)
             next_token = next(tokens, None)
 
         if (

--- a/sqlvalidator/grammar/sql.py
+++ b/sqlvalidator/grammar/sql.py
@@ -1557,6 +1557,12 @@ class ExceptClause(SelectAllClause):
 class ReplaceClause(SelectAllClause):
     KEYWORD = "REPLACE"
 
+    def validate(self, known_fields: Set[_FieldInfo]) -> list:
+        errors = super().validate(known_fields)
+        for arg in self.args:
+            errors += arg.alias.validate(known_fields)
+        return errors
+
 
 class Case(Expression):
     def __init__(self, expression, when_then, else_expression):

--- a/sqlvalidator/grammar/sql.py
+++ b/sqlvalidator/grammar/sql.py
@@ -1548,6 +1548,25 @@ class ExceptClause(Expression):
         return except_str + ")"
 
 
+class ReplaceClause(Expression):
+    def __init__(self, expression, args):
+        super().__init__(expression)
+        self.args = args
+
+    def __repr__(self):
+        return "<{}: {!r} args: {!r}>".format(
+            self.__class__.__name__, self.value, self.args
+        )
+
+    def __str__(self):
+        except_str = "{} REPLACE (".format(transform(self.value))
+        if len(self.args) > 1:
+            except_str += "\n {}\n".format(",\n ".join(map(transform, self.args)))
+        elif except_str:
+            except_str += transform(self.args[0])
+        return except_str + ")"
+
+
 class Case(Expression):
     def __init__(self, expression, when_then, else_expression):
         super().__init__(expression)

--- a/sqlvalidator/grammar/sql.py
+++ b/sqlvalidator/grammar/sql.py
@@ -1529,7 +1529,9 @@ class Negation(Expression):
         return bool
 
 
-class ExceptClause(Expression):
+class SelectAllClause(Expression):
+    KEYWORD = ""
+
     def __init__(self, expression, args):
         super().__init__(expression)
         self.args = args
@@ -1540,7 +1542,7 @@ class ExceptClause(Expression):
         )
 
     def __str__(self):
-        except_str = "{} EXCEPT (".format(transform(self.value))
+        except_str = "{} {} (".format(transform(self.value), self.KEYWORD)
         if len(self.args) > 1:
             except_str += "\n {}\n".format(",\n ".join(map(transform, self.args)))
         elif except_str:
@@ -1548,23 +1550,12 @@ class ExceptClause(Expression):
         return except_str + ")"
 
 
-class ReplaceClause(Expression):
-    def __init__(self, expression, args):
-        super().__init__(expression)
-        self.args = args
+class ExceptClause(SelectAllClause):
+    KEYWORD = "EXCEPT"
 
-    def __repr__(self):
-        return "<{}: {!r} args: {!r}>".format(
-            self.__class__.__name__, self.value, self.args
-        )
 
-    def __str__(self):
-        except_str = "{} REPLACE (".format(transform(self.value))
-        if len(self.args) > 1:
-            except_str += "\n {}\n".format(",\n ".join(map(transform, self.args)))
-        elif except_str:
-            except_str += transform(self.args[0])
-        return except_str + ")"
+class ReplaceClause(SelectAllClause):
+    KEYWORD = "REPLACE"
 
 
 class Case(Expression):

--- a/tests/integration/test_formatting.py
+++ b/tests/integration/test_formatting.py
@@ -2152,3 +2152,33 @@ SELECT col = "Test\'"
 FROM t
 """
     assert format_sql(sql) == expected.strip()
+
+
+def test_select_star_replace():
+    sql = 'select * replace ("x" as col) from t'
+    expected = """
+SELECT * REPLACE ("x" AS col)
+FROM t
+"""
+    assert format_sql(sql) == expected.strip()
+
+
+def test_select_star_multiple_replace():
+    sql = 'select * replace ("x" as col, "y" as f) from t'
+    expected = """
+SELECT * REPLACE (
+ "x" AS col,
+ "y" AS f
+)
+FROM t
+"""
+    assert format_sql(sql) == expected.strip()
+
+
+def test_select_star_replace_nospace():
+    sql = 'select * replace("x" as col) from t'
+    expected = """
+SELECT * REPLACE ("x" AS col)
+FROM t
+"""
+    assert format_sql(sql) == expected.strip()

--- a/tests/integration/test_validation.py
+++ b/tests/integration/test_validation.py
@@ -545,3 +545,19 @@ def test_between_condition_is_valid():
       f BETWEEN 200 AND 299
     """
     assert_valid_sql(sql)
+
+
+def test_select_star_replace():
+    sql = """
+SELECT * REPLACE ("x" AS col)
+FROM t
+"""
+    assert_valid_sql(sql)
+
+
+def test_select_star_unknown_column():
+    sql = """
+SELECT * REPLACE ("x" AS col)
+FROM (select field from t)
+"""
+    assert_invalid_sql(sql)

